tiledb/sm/filter/xor_filter.cc

```diff
@@ -0,0 +1,248 @@
+ *
+ * This file implements class XORFilter.
+ */
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/enums/filter_option.h"
+#include "tiledb/sm/filter/filter_buffer.h"
+#include "tiledb/sm/tile/tile.h"
+
+#include "tiledb/sm/filter/xor_filter.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+void XORFilter::dump(FILE* out) const {
+  if (out == nullptr)
+    out = stdout;
+  fprintf(out, "XORFilter");
+}
+
+Status XORFilter::run_forward(
+    const Tile& tile,
+    Tile* const,  // offsets_tile
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  auto tile_type = tile.type();
+
+  switch (tile_type) {
+    case Datatype::INT8: {
+      return run_forward<int8_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case Datatype::INT16: {
+      return run_forward<int16_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case Datatype::INT32: {
+      return run_forward<int32_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case Datatype::INT64: {
+      return run_forward<int64_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    default: {
+      return Status_FilterError("XORFilter::run_forward: invalid datatype.");
+    }
+  }
+
+  return Status::Ok();
+}
+
+template <typename T>
+Status XORFilter::run_forward(
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  // Output size does not change with this filter.
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Write the metadata
+  auto parts = input->buffers();
+  auto num_parts = (uint32_t)parts.size();
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // Shuffle all parts
+  for (const auto& part : parts) {
+    auto part_size = (uint32_t)part.size();
+    RETURN_NOT_OK(output_metadata->write(&part_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(shuffle_part<T>(&part, output_buf));
+  }
+
+  return Status::Ok();
+}
+
+template <typename T>
+Status XORFilter::shuffle_part(const ConstBuffer* part, Buffer* output) const {
+  uint32_t s = part->size();
+  assert(s % sizeof(T) == 0);
+  uint32_t num_elems_in_part = s / sizeof(T);
+
+  if (num_elems_in_part == 0) {
+    return Status::Ok();
+  }
+
+  // Write starting element.
+  const T* part_array = static_cast<const T*>(part->data());
+  RETURN_NOT_OK(output->write(part_array, sizeof(T)));
+  output->advance_offset(sizeof(T));
+
+  for (uint32_t j = 1; j < num_elems_in_part; ++j) {
+    T value = part_array[j] - part_array[j - 1];
```

Comment:
Sorry for the late comment, and maybe it's obvious, but I am having trouble to understand why we consider substraction of `T` types to be `XOR` equivalent, since substraction can have borrows. Could you please explain the rationale behind e.g the mathematical identity or the reason that in our case they are equivalent so that I understand better? 
I already checked the design document of `LIDAR` compressor but I didn't find the explanation for this assumption there either.

Reply:
Honestly good question, I have no idea and thanks for bringing this up. @ihnorton or @stavrospapadopoulos ?

Reply:
This has been clarified, feel free to leave a review soonish! @ypatia

Reply:
Thanks Abi, can you share what is the conclusion of the clarification though?

Reply:
By looking at the code I guess it's that it should be XOR operation instead of substraction?

Reply:
yes, it should be

Reply:
Thanks Abi, my concern is now addressed, so it's no objection on my side for the PR. However, I see that other reviewers still have comments so I'll let them approve when they are addressed.